pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -45,31 +45,46 @@ import (
 	"k8s.io/klog"
 )
 
+type configGetter func(v1alpha1.ProviderSpec) (*Config, *providerconfigtypes.Config, *vspheretypes.RawConfig, error)
+
+// Ensures that getConfig function is of type configGetter
+var _ configGetter = (&provider{}).getConfig
+
 type provider struct {
 	configVarResolver *providerconfig.ConfigVarResolver
+	// This is to ease unit tests
+	getConfigFunc configGetter
```

Comment:
Why this is needed? there is already a method which get the configurations of the provider for you.

Reply:
Instead of defining a new unneeded field to the provider to test something, you can use the getConfig in your tests and pass your custom configurations there instead of finding a way to override those values.

Reply:
It's to ease unit tests, I'm mocking this function in the unit test:
https://github.com/irozzo-1A/machine-controller/blob/support-datastore-cluster/pkg/cloudprovider/provider/vsphere/provider_test.go#L136

Reply:
Yeah as I said this you can simply use `v1alpha1.ProviderSpec` object to customize your test. I am not sure it is a good idea to add code that belongs to the tests in the production code that code should be built in the matter that its testable and it is in this case.

Reply:
> Instead of defining a new unneeded field to the provider to test something, you can use the getConfig in your tests and pass your custom configurations there instead of finding a way to override those values.

I know that I can but it's not the `getConfig` method that I want to test here. I don't want to open a philosophical debate whether changing the code to ease testing is a good practice or not, but I normally prefer to reduce the scope of unit test as much as I can. I found this pattern somewhere in the k8s code base quite handy. What's your concern about adding an unexported field?

Reply:
Well, this field is simply not needed and the logic behind can be substituted by using a method that already exists. I know that you are not testing the `getConfig` method but you want to test something that the `getConfig` assigns in the runtime code. All our cloud providers are following the very same pattern.

Reply:
to be fair the other providers have no unit tests, then the problem does not arise :-)
There are many ways of doing things, in absolute terms it's not required, but it's functional to its purpose. It simplifies a bit the test at the cost of complexifying a bit the code. I think it's largely matters of taste. We can discuss pros and cons offline if you wish, that said I think it's not worthy spending much time on this point, if it's a blocker for you I can invest some time to change it.

Reply:
(IMHO) Adding code in order to facilitate your test is a bad design. The reason for that is, most of the time when code is added in production only to facilitate/enable testing show some design flaws, which is funny in your case because you can test the code without adding anything new or adding any wrapping. You can easily test your code without adding anything, however, you simply chose to:

- Add something in the production code which has no value 
- Ignore what's already there and can be used and reflects the applications' workflow in production.
- Breaking the code consistency between this particular cloud provider and the others without realistic and valuable reasons.
 
If none of those points are sufficient we can take this discussion as you suggested offline.

Reply:
@moadqassem thanks for clarifying your view, I honestly think that your remark is more aesthetic than substantial, but as you dislike it so much and removing it does not cost me much effort I'll remove it ;-)

Reply:
done